substrate/frame/support/procedural/src/lib.rs

```diff
@@ -1506,18 +1506,81 @@ pub fn origin(_: TokenStream, _: TokenStream) -> TokenStream {
 ///
 /// ```ignore
 /// Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, MaxEncodedLen, TypeInfo,
-/// RuntimeDebug
+/// RuntimeDebug,
+/// CompositeEnumVariantCount,
 /// ```
 ///
 /// For ease of usage, when no `#[derive]` attributes are found for the enum under
 /// `#[pallet::composite_enum]`, the aforementioned traits are automatically derived for it. The
 /// inverse is also true: if there are any `#[derive]` attributes found for the enum, then no traits
-/// will automatically be derived for it.
+/// will automatically be derived for it (this implies that you need to provide the
+/// `frame_support::traits::VariantCount` implementation).
 #[proc_macro_attribute]
 pub fn composite_enum(_: TokenStream, _: TokenStream) -> TokenStream {
 	pallet_macro_stub()
 }
 
+/// Derive macro for implementing `frame_support::traits::VariantCount` for enum type.
+///
+/// This macro works only for the enum which does not have variants or variants do not have fields.
+///
+/// This macro is automatically derived for `#[pallet::composite_enum]` enums.
+#[proc_macro_derive(CompositeEnumVariantCount)]
```

Comment:
Honestly, I have never created any derive macro before, and it took me a while to figure out how `#[pallet::composite_enum]` with `construct_runtime` works and is evaluated :)

If you mean [this](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/support/procedural/src/lib.rs#L1494-L1519) by `composite macro directly`, I don't know, iiuc this is just a placeholder for the case that anybody mistakenly uses `#[composite_enum]` instead of `#[pallet::composite_enum]`.

I don't see any expand stuff for `composites`, the only place I can see is in [`CompositeDef::try_from`](https://github.com/paritytech/polkadot-sdk/blob/bko-max-holds-integrity-test/substrate/frame/support/procedural/src/pallet/parse/composite.rs#L134-L145) where derive marcos are added, so that's why I created this one and also added there.

I can see that `#[pallet::*]` stuff is expanded [here](https://github.com/paritytech/polkadot-sdk/blob/bko-max-holds-integrity-test/substrate/frame/support/procedural/src/pallet/expand/mod.rs#L57-L78) but `composites` are not used here.


@bkchr I don't know, what do you suggest here?

Reply:
@bkontur This can be added as a separate call during the pallet expansion as [linked](https://github.com/paritytech/polkadot-sdk/blob/bko-max-holds-integrity-test/substrate/frame/support/procedural/src/pallet/expand/mod.rs#L57-L78) by you already.

Reply:
> @bkontur This can be added as a separate call during the pallet expansion as [linked](https://github.com/paritytech/polkadot-sdk/blob/bko-max-holds-integrity-test/substrate/frame/support/procedural/src/pallet/expand/mod.rs#L57-L78) by you already.

@gupnik ok, thank you, no problem, I will do that. Off the record, I just want to understand because you both suggest the same thing and as I said bellow this was my first derive macro ever :). Could you tell me the rationale behind **not having a derive macro** versus **doing it during pallet expansions**? Is there any downside to adding a derive macro here?

Reply:
@bkontur I find that this macro assumes a number of things regarding the site where its applied as can be seen from your comments throughout - essentially being specific to the composite enum as defined elsewhere. So, I am not sure if having it a separate entity really adds much value here.